### PR TITLE
Add explcitly testing of deep_copy with scalar type conversion

### PR DESCRIPTION
**Pull request by crtrott**

### CODE REVIEW

File: core/unit_test/TestDeepCopy.hpp [hunk 1]
```diff
@@ -210,5 +211,147 @@ TEST(TEST_CATEGORY, deep_copy_alignment) {
...
+
+    Kokkos::deep_copy(view_s1_1d, static_cast<Scalar1>(0));
+    Kokkos::deep_copy(view_s1_2d, static_cast<Scalar1>(0));
+
+    Kokkos::parallel_for("TestDeepCopyConversion::Compare",
+                         Kokkos::RangePolicy<TEST_EXECSPACE, TagCompare,
+                                             Kokkos::IndexType<int64_t>>(0, N0),
+                         *this);
+    Kokkos::deep_copy(errors, error_count);
+    ASSERT_TRUE(errors > 0);
```
Comment (Rombur) [inline core/unit_test/TestDeepCopy.hpp]: HIP is not happy about this check.

File: core/unit_test/TestDeepCopy.hpp [hunk 2]
```diff
@@ -210,5 +211,147 @@ TEST(TEST_CATEGORY, deep_copy_alignment) {
...
+  void operator()(TagCompare, const int64_t i) const {
+    int64_t errors = 0;
+    if (view_s1_1d(i) != static_cast<Scalar1>(static_cast<Scalar2>(i + 1)))
+      errors++;
+    for (int64_t j = 0; j < static_cast<int64_t>(view_s1_2d.extent(1)); j++) {
+      if (view_s1_2d(i, j) !=
+          static_cast<Scalar1>(static_cast<Scalar2>((i + 1) * 1000 + j + 1)))
+        errors++;
+    }
+    if (errors > 0) Kokkos::atomic_add(&error_count(), errors);
```
Comment (Rombur) [inline core/unit_test/TestDeepCopy.hpp]: Looks like there is a problem with the atomic_add with HIP, `errors` has the correct value. We haven't really worked on the `atomic_add` in HIP. So it's not surprising that it doesn't work.

Comment (crtrott) [reply core/unit_test/TestDeepCopy.hpp]: Ok gonna disable the test for HIP for now.

File: core/unit_test/TestDeepCopy.hpp [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 namespace Test {
 
+#ifdef KOKKOS_ENABLE_CXX11_DISPATCH_LAMBDA
```
Comment (DavidPoliakoff) [inline core/unit_test/TestDeepCopy.hpp]: This feels orthogonal to the intent of the PR. I'll still merge, but want to make sure I'm not missing some implementation detail

Comment (DavidPoliakoff) [reply core/unit_test/TestDeepCopy.hpp]: Answered via Slack, this was moved from where the test was included to surround the test. That's a lot more informative to me

**Pull request merged**